src/index/index.c

```diff
@@ -1,173 +1,316 @@
-  skiplistDelete(sl, val, &node);
+    RediSearch_SpecAddDocument(rsIdx, doc);
 }
- void Index_InsertNode(Index *idx, NodeID node, SIValue *val) {
-  skiplist *sl = _select_skiplist(idx, val->type);
-  if (!sl) return; // Value was of a type not supported by indices.
-  skiplistInsert(sl, val, node);
+
+void Index_RemoveNode
+(
+    Index *idx,   // Index to use
+    Node *n         // Node to remove
+)
+{
+    assert(idx && n);
+    NodeID node_id = ENTITY_GET_ID(n);
+    RediSearch_DeleteDocument(idx->idx, &node_id, sizeof(EntityID));
 }
 
-//------------------------------------------------------------------------------
-// Index iterator functions
-//------------------------------------------------------------------------------
+// Constructs index.
+void Index_Construct
+(
+    Index *idx
+)
+{
+    assert(idx);
+    
+    /* RediSearch index already exists
+     * re-construct */
+    if(idx->idx) RediSearch_DropIndex(idx->idx);
+
+    RSIndex *rsIdx = NULL;
+    RSIndexOptions *idx_options = RediSearch_CreateIndexOptions();
+    RediSearch_IndexOptionsSetGetValueCallback(idx_options, _getNodeAttribute, NULL);
+    rsIdx = RediSearch_CreateIndex(idx->label, idx_options);
+    RediSearch_FreeIndexOptions(idx_options);
+
+    // Create indexed fields
+    if(idx->type == IDX_FULLTEXT) {
+        for(uint i = 0; i < idx->fields_count; i++) {
+            // Introduce text field.
+            RediSearch_CreateTextField(rsIdx, idx->fields[i]);
+        }
+    } else {
+        char buff[1024];
+        for(uint i = 0; i < idx->fields_count; i++) {
+            // Introduce both text and numeric fields.
+            sprintf(buff, "%s_tag", idx->fields[i]);
+            RediSearch_CreateTagField(rsIdx, buff);
+            sprintf(buff, "%s_numeric", idx->fields[i]);
+            RediSearch_CreateNumericField(rsIdx, buff);
+        }
+    }
+
+    idx->idx = rsIdx;
 
-/* Generate an iterator with no lower or upper bound. */
-IndexIter* IndexIter_Create(Index *idx, SIType type) {
-  skiplist *sl = (type == T_STRING) ? idx->string_sl : idx->numeric_sl;
-  return skiplistIterateAll(sl);
+    _populateIndex(idx);
 }
 
-/* Apply a filter to an iterator, modifying the appropriate bound if
- * it narrows the iterator range.
- * Returns 1 if the filter was a comparison type that can be translated into a bound
- * (effectively, any type but '!='), which indicates that it is now redundant. */
-bool IndexIter_ApplyBound(IndexIter *iter, SIValue *bound, int op) {
-  return skiplistIter_UpdateBound(iter, bound, op);
+// Query index.
+RSResultsIterator* Index_Query
+(
+    const Index *idx,
+    const char *query,           // Query to execute
+    char** err
+)
+{
+    assert(idx && query);
+    return RediSearch_IterateQuery(idx->idx, query, strlen(query), err);
 }
 
-NodeID* IndexIter_Next(IndexIter *iter) {
-  return skiplistIterator_Next(iter);
+// Return indexed label.
+char* Index_GetLabel
```

Comment:
Unused. If you want to keep it for the interface, I'd not strdup and return `const`.

Reply:
Used by test_index.cpp

Reply:
changed to const, remove string duplication